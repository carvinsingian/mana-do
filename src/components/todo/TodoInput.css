.ToDoInput__container {
  flex: 1 1;
}

.ToDoInput__input {
  width: 100%;
}
