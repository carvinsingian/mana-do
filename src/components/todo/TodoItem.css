.ToDoItem__container {
  display: flex;
  align-items: center;
  justify-content: space-between;
  border-bottom: 1px solid gray;
  margin-top: 4px;
}

.ToDoItem__container > span {
  flex: 1 1;
  text-align: left;
  margin-left: 8px;
}

.ToDoItem__input {
  width: 100%;
}

.ToDoItem__input--completed {
  text-decoration: line-through;
  opacity: 0.5;
}

.ToDoItem__delete {
  background-color: transparent;
  color: white;
  border: none;
  width: 32px;
  height: 32px;
  min-width: auto;
  min-height: auto;
  box-shadow: none;
  border-radius: 50%;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
}
