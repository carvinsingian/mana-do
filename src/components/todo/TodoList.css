.ToDoList__container {
  display: flex;
  flex-direction: column;
}

.ToDoList__header {
  margin-top: 1.5rem;
}

.ToDoList__text {
  text-align: center;
  margin-top: 8px;
}
