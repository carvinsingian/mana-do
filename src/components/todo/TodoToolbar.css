.ToDoToolbar__container {
  margin-top: 16px;
  width: 100%;
}

/* .ToDoToolbar__tabs {
  display: flex;
  justify-content: center;
  align-items: center;
}

.ToDoToolbar__tabs--active {
  background-color: darkcyan;
}

.ToDoToolbar__tabs > *:not(:last-child) {
  margin-right: 8px;
} */

.ToDoToolbar__actions {
  display: flex;
  margin-top: 16px;
  align-items: center;
  justify-content: space-between;
}

.ToDoToolbar__actions--hide {
  height: 48px;
  width: fit-content;
  margin-left: 8px;
}

.ToDoToolbar__actions--delete {
  height: 48px;
  width: fit-content;
  margin-left: 8px;
  background-color: rgb(211, 35, 35);
}

.ToDoToolbar__actions--active {
  background-color: darkcyan;
}
