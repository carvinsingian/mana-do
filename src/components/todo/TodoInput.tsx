import React, { useRef } from 'react'
import './TodoInput.css'

interface Props {
  onCreateTodo: (todo: string) => Promise<void>
}

const TodoInput = ({ onCreateTodo }: Props) => {
  const inputRef = useRef<HTMLInputElement>(null)

  const handleCreateTodo = (e: React.KeyboardEvent<HTMLInputElement>) => {
    //Ignore enter key when there is no value
    if (e.key === 'Enter' && inputRef.current && inputRef.current.value !== '') {
      onCreateTodo(inputRef.current.value)
      inputRef.current.value = '';
    }
  }

  return (
    <div className="ToDoInput__container">
      <input
        ref={inputRef}
        data-testid="todo-input"
        className="ToDoInput__input"
        placeholder="What needs to be done?"
        onKeyDown={handleCreateTodo}
      />
    </div>
  )
}

export default TodoInput