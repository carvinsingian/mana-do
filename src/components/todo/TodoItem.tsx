import React from 'react'

import { isTodoCompleted } from '../../utils';
import {Todo } from '../../models/todo';
import './TodoItem.css'

interface Props {
  todo: Todo
  editInputRef: React.RefObject<HTMLInputElement>
  editTodoItem: Todo | null
  setEditTodoItem: (todo: Todo | null) => void
  onEditTodo: (todo: Todo) => Promise<void>
  onUpdateTodoStatus: (e: React.ChangeEvent<HTMLInputElement>, todoId: string) => void
  onDeleteTodo: (todo: Todo) => void
}

const TodoItem = ({ todo, editInputRef, editTodoItem, setEditTodoItem, onEditTodo, onUpdateTodoStatus, onDeleteTodo } : Props) => {
  const handleEditTodo = (e: React.KeyboardEvent<HTMLInputElement>) => {
    //Ignore enter key when there is no value
    if (e.key === 'Enter' && editTodoItem && editTodoItem.content !== '') {
      onEditTodo(editTodoItem)
      setEditTodoItem(null)
    }
  }

  return (
    <div className="ToDoItem__container">
      <input
          data-testid='todo-item-checkbox'
          type="checkbox"
          checked={isTodoCompleted(todo)}
          onChange={(e) => onUpdateTodoStatus(e, todo.id)}
      />
      {/* Add a condition to display input when there is an edit item and it's equal to todo item id */}
      {editTodoItem && editTodoItem.id === todo.id ? (
        <input
          data-testid='todo-item-edit'
          ref={editInputRef}
          className='ToDoItem__input'
          value={editTodoItem.content}
          onKeyDown={handleEditTodo}
          onChange={(e) => {
            setEditTodoItem({...todo, content: e.target.value})
          }}
        />
      ) : (
        <span data-testid='todo-item-span' className={`${isTodoCompleted(todo) ? 'ToDoItem__input--completed' : ''}`}onDoubleClick={() => setEditTodoItem(todo)}>{todo.content}</span>
      )}
      
      <button
        data-testid="todo-item-delete"
        className="ToDoItem__delete"
        onClick={() => onDeleteTodo(todo)}
      >
        delete
      </button>
    </div>
  )
}


export default TodoItem