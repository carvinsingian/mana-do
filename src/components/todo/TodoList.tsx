import React, { useEffect, useRef, useState } from 'react'
import { Todo } from '../../models/todo';
import useOutsideClickHook from '../../hooks/OutsideClickHook';
import TodoItem from './TodoItem';
import './TodoList.css'
import { filterTodos } from '../../utils'

interface Props {
  todos: Todo[]
  isShowAll: boolean
  onEditTodo: (todo: Todo) => Promise<void>
  onUpdateTodoStatus: (e: React.ChangeEvent<HTMLInputElement>, todoId: string) => void
  onDeleteTodo: (todo: Todo) => void
}

const TodoList = ({ todos, isShowAll, onEditTodo, onUpdateTodoStatus, onDeleteTodo } : Props) => {
  const [editTodoItem, setEditTodoItem] = useState<Todo | null>(null);

  //Create editInputReference to use for determining if the user is clicking outside
  const editInputRef = useRef<HTMLInputElement>(null);
  const isCloseEdit = useOutsideClickHook(editInputRef)

  const { activeTodos, completedTodos } = filterTodos(todos)

  useEffect(() => {
    if (isCloseEdit) {
      setEditTodoItem(null)
    }
  }, [isCloseEdit])

  const getTodoItems = (todos: Todo[]) => {
    return (
      todos.map((todo, index) => {
        return (
          <TodoItem 
            key={index} 
            todo={todo} 
            editInputRef={editInputRef} 
            editTodoItem={editTodoItem}
            setEditTodoItem={setEditTodoItem}
            onEditTodo={onEditTodo}
            onUpdateTodoStatus={onUpdateTodoStatus}
            onDeleteTodo={onDeleteTodo}
          />
        );
      })
    )
  }

  return (
    <div className="ToDoList__container">
      <div className="ToDoList__header">
        <h2>Active Tasks</h2>
      </div>
      { activeTodos.length > 0 ? ( 
          getTodoItems(activeTodos)
        ) : (
          <div className="ToDoList__text">No Todo Tasks</div>
        )
      }
      { isShowAll && (
          <div className="ToDoList__header">
            <h2>Completed Tasks</h2>
          </div>
        )
      }
      { isShowAll && (
          completedTodos.length > 0 ? (
            getTodoItems(completedTodos)
          ) : (
            <div className="ToDoList__text">No Todo Tasks</div>
          )
      )}
    </div>
  )
}

export default TodoList