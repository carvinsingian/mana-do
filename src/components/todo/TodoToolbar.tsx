
import React from 'react'
import Button from '../../ui/button/Button';
import { Todo } from '../../models/todo';
import './TodoToolbar.css'


interface Props {
  todos: Todo[]
  isShowAll: boolean
  onToggleAllTodo: (e: React.ChangeEvent<HTMLInputElement>) => void
  onDeleteAllTodo: () => void
  onToggleCompletedTodo: () => void
}

const TodoToolbar = ({ todos, isShowAll, onToggleAllTodo, onDeleteAllTodo, onToggleCompletedTodo }: Props) => {

  const isTodos = () => {
    return todos.length > 0
  }

  return (
    <div className="ToDoToolbar__container">
      {isTodos() && (
        <div className="ToDoToolbar__actions">
          <input
            data-testid='todo-toolbar-checkbox'
            type="checkbox"
            onChange={onToggleAllTodo}
          />
            <Button
              dataTestId='todo-toolbar-complete-button'
              className={`ToDoToolbar__actions--hide ${isShowAll ? 'ToDoToolbar__actions--active' : ''}`} 
              variant="primary" 
              onClick={onToggleCompletedTodo}
            >
              Completed Tasks
            </Button>
          <Button
            dataTestId='todo-toolbar-delete-button'
            className="ToDoToolbar__actions--delete" 
            variant="primary" 
            onClick={onDeleteAllTodo}
          >
            Clear all todos
          </Button>
        </div>
      )}

    </div>
  )
}

export default TodoToolbar