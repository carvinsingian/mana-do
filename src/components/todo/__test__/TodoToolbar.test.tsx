import React from 'react'
import ReactDOM from 'react-dom'
import TodoToolbar from '../TodoToolbar'
import { render, cleanup } from '@testing-library/react'
import { TodoStatus } from '../../../models/todo'

afterEach(cleanup)

test('Todo toolbar renders without crashing', () => {
  const div = document.createElement('div')
  const todo = {
    id: 'testid',
    user_id: 'userid',
    content : 'content',
    status: TodoStatus.ACTIVE,
    created_date: 'date'
  }
  const isShowAll= true
  const onToggleAllTodo = jest.fn()
  const onDeleteAllTodo = jest.fn()
  const onToggleCompletedTodo = jest.fn()

  ReactDOM.render(
  <TodoToolbar
    todos={[todo]}
    isShowAll={isShowAll}
    onToggleAllTodo={onToggleAllTodo}
    onDeleteAllTodo={onDeleteAllTodo}
    onToggleCompletedTodo={onToggleCompletedTodo}
  />, div)
})

test('Todo toolbar matches snapshots with todos', () => {
  const todo = {
    id: 'testid',
    user_id: 'userid',
    content : 'content',
    status: TodoStatus.ACTIVE,
    created_date: 'date'
  }
  const isShowAll= true
  const onToggleAllTodo = jest.fn()
  const onDeleteAllTodo = jest.fn()
  const onToggleCompletedTodo = jest.fn()

  const { asFragment } = render(
    <TodoToolbar
      todos={[todo]}
      isShowAll={isShowAll}
      onToggleAllTodo={onToggleAllTodo}
      onDeleteAllTodo={onDeleteAllTodo}
      onToggleCompletedTodo={onToggleCompletedTodo}
    />
  )
  const componentRender = asFragment()
  expect(componentRender).toMatchSnapshot()
})

test('Todo toolbar matches snapshots without todos', () => {
  const isShowAll= true
  const onToggleAllTodo = jest.fn()
  const onDeleteAllTodo = jest.fn()
  const onToggleCompletedTodo = jest.fn()

  const { asFragment } = render(
    <TodoToolbar
      todos={[]}
      isShowAll={isShowAll}
      onToggleAllTodo={onToggleAllTodo}
      onDeleteAllTodo={onDeleteAllTodo}
      onToggleCompletedTodo={onToggleCompletedTodo}
    />
  )
  const componentRender = asFragment()
  expect(componentRender).toMatchSnapshot()
})

test('Todo toolbar input and buttons are present when there are todos', () => {
  const todo = {
    id: 'testid',
    user_id: 'userid',
    content : 'content',
    status: TodoStatus.ACTIVE,
    created_date: 'date'
  }
  const isShowAll= true
  const onToggleAllTodo = jest.fn()
  const onDeleteAllTodo = jest.fn()
  const onToggleCompletedTodo = jest.fn()

  const { getByTestId } = render(
    <TodoToolbar
      todos={[todo]}
      isShowAll={isShowAll}
      onToggleAllTodo={onToggleAllTodo}
      onDeleteAllTodo={onDeleteAllTodo}
      onToggleCompletedTodo={onToggleCompletedTodo}
    />
  )

  const todoToolbarCheckbox = getByTestId('todo-toolbar-checkbox')
  const todoToolbarComplete = getByTestId('todo-toolbar-complete-button')
  const todoToolbarDelete = getByTestId('todo-toolbar-delete-button')

  expect(todoToolbarCheckbox).toBeInTheDocument()
  expect(todoToolbarComplete).toBeInTheDocument()
  expect(todoToolbarDelete).toBeInTheDocument()
})

