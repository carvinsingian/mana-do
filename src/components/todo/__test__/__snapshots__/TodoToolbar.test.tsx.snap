// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo toolbar matches snapshots with todos 1`] = `
<DocumentFragment>
  <div
    class="ToDoToolbar__container"
  >
    <div
      class="ToDoToolbar__actions"
    >
      <input
        data-testid="todo-toolbar-checkbox"
        type="checkbox"
      />
      <button
        class="ToDoToolbar__actions--hide ToDoToolbar__actions--active button button__primary"
        data-testid="todo-toolbar-complete-button"
        type="button"
      >
        Completed Tasks
      </button>
      <button
        class="ToDoToolbar__actions--delete button button__primary"
        data-testid="todo-toolbar-delete-button"
        type="button"
      >
        Clear all todos
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Todo toolbar matches snapshots without todos 1`] = `
<DocumentFragment>
  <div
    class="ToDoToolbar__container"
  />
</DocumentFragment>
`;
